File: ios/MenCare/Features/Admin/Dashboard/DashboardScreen.swift
import SwiftUI
import Observation

/// Il cruscotto leggeva il repository dentro `body`: con una chiamata di rete
/// non si puo' piu', e i numeri li calcola il server. Qui c'e' il view model che
/// mancava.
@MainActor
@Observable
final class DashboardViewModel {

    private let admin: AdminRepository

    let state = LoadState()
    var period: DashboardPeriod = .day
    private(set) var stats: DashboardStats?

    init(admin: AdminRepository) {
        self.admin = admin
    }

    func load() async {
        await state.run { stats = try await admin.dashboard(period) }
    }

    func select(_ newPeriod: DashboardPeriod) {
        guard newPeriod != period else { return }
        period = newPeriod
        Task { await load() }
    }
}

struct DashboardScreen: View {
    @State var viewModel: DashboardViewModel
    let onSendCampaign: () -> Void

    private var period: DashboardPeriod { viewModel.period }

    var body: some View {
        Group {
            if let stats = viewModel.stats {
                content(stats)
            } else if let error = viewModel.state.error {
                VStack(spacing: 0) {
                    periodHeader
                    BrandErrorRetry(error: error, retry: { Task { await viewModel.load() } })
                    Spacer()
                }
            } else {
                VStack(spacing: 0) {
                    periodHeader
                    BrandLoading()
                    Spacer()
                }
            }
        }
        .background(Color.bone)
        .inkStatusBarBackdrop()
        .task { await viewModel.load() }
    }

    /// La banda con i tre periodi resta anche mentre si carica: cambiare
    /// periodo e' l'unica cosa da fare se la lettura sta fallendo.
    private var periodHeader: some View {
        DarkHeader(
            roundedBottom: true,
            contentPadding: EdgeInsets(top: 8, leading: 20, bottom: 18, trailing: 20)
        ) {
            // La banda porta il suo titolo, come le altre schermate del
            // titolare: prima si sa dove si è, poi si sceglie il periodo.
            Text(L("dash_overline").uppercased())
                .font(Typo.overline)
                .foregroundStyle(Color.oliveLight)
            Text(L("dash_title"))
                .font(Typo.displaySmall)
                .foregroundStyle(Color.bone)
                .padding(.top, 4)
                .padding(.bottom, 16)
            SegmentedTabs(
                options: [L("dash_period_day"), L("dash_period_week"), L("dash_period_month")],
                selectedIndex: DashboardPeriod.allCases.firstIndex(of: period) ?? 0,
                onSelect: { viewModel.select(DashboardPeriod.allCases[$0]) }
            )
        }
    }

    private func content(_ stats: DashboardStats) -> some View {
        ScrollView {
            VStack(spacing: 0) {
                if let error = viewModel.state.error {
                    InlineErrorBar(message: error.displayMessage, retry: { Task { await viewModel.load() } })
                }
                DarkHeader(
                    roundedBottom: true,
                    contentPadding: EdgeInsets(top: 8, leading: 20, bottom: 18, trailing: 20)
                ) {
                    // La banda porta il suo titolo, come le altre schermate del
                    // titolare: prima si sa dove si è, poi si sceglie il periodo.
                    Text(L("dash_overline").uppercased())
                        .font(Typo.overline)
                        .foregroundStyle(Color.oliveLight)
                    Text(L("dash_title"))
                        .font(Typo.displaySmall)
                        .foregroundStyle(Color.bone)
                        .padding(.top, 4)
                        .padding(.bottom, 16)
                    SegmentedTabs(
                        options: [L("dash_period_day"), L("dash_period_week"), L("dash_period_month")],
                        selectedIndex: DashboardPeriod.allCases.firstIndex(of: period) ?? 0,
                        onSelect: { viewModel.select(DashboardPeriod.allCases[$0]) }
                    )
                    Text(revenueLabel.uppercased())
                        .font(Typo.jost(11, weight: .medium))
                        .kerning(1.6)
                        .foregroundStyle(Color.oliveWood)
                        .padding(.top, 16)
                    HStack {
                        Text(formatPriceCompact(stats.revenueCents))
                            .font(Typo.cormorant(44))
                            .foregroundStyle(Color.bone)
                        Spacer()
                        Text(L("dash_trend", stats.revenueTrendPercent))
                            .font(Typo.titleSmall)
                            .foregroundStyle(Color.bone)
                            .padding(.horizontal, 12)
                            .padding(.vertical, 7)
                            .background(RoundedRectangle(cornerRadius: 10).fill(Color.oliveWood))
                    }
                    revenueBars
                        .padding(.top, 10)
                }

                VStack(alignment: .leading, spacing: 12) {
                    HStack(spacing: 10) {
                        // I tre KPI hanno lo stesso aspetto: fondo chiaro e filo, niente grigio.
                        StatTile(
                            label: L("dash_appointments"),
                            value: "\(stats.appointmentCount)",
                            container: .bone,
                            borderColor: .stoneBorder
                        )
                        StatTile(
                            label: L("dash_no_show"),
                            value: String(format: "%.1f%%", stats.noShowPercent).replacingOccurrences(of: ".", with: ","),
                            container: .bone,
                            borderColor: .stoneBorder
                        )
                        StatTile(
                            label: L("dash_avg_ticket"),
                            value: formatPriceCompact(stats.averageTicketCents),
                            container: .bone,
                            borderColor: .stoneBorder
                        )
                    }
                    .fixedSize(horizontal: false, vertical: true)

                    VStack(alignment: .leading, spacing: 9) {
                        BrandSectionLabel(text: L("dash_operator_occupancy"))
                        ForEach(stats.operatorOccupancy) { row in
                            HStack(spacing: 10) {
                                Text(String(row.operatorName.split(separator: " ").first ?? ""))
                                    .font(Typo.bodyMedium)
                                    .foregroundStyle(Color.ink)
                                    .frame(width: 64, alignment: .leading)
                                GeometryReader { geo in
                                    ZStack(alignment: .leading) {
                                        Capsule().fill(Color.stone)
                                        Capsule()
                                            .fill(row.percent >= 80 ? Color.oliveWood : Color.ink)
                                            .frame(width: geo.size.width * CGFloat(row.percent) / 100)
                                    }
                                }
                                .frame(height: 7)
                                Text("\(row.percent)%")
                                    .font(Typo.titleSmall)
                                    .foregroundStyle(Color.ink)
                                    .frame(width: 44, alignment: .trailing)
                            }
                        }
                    }

                    upcomingSection(stats.upcomingDays)
                }
                .padding(.horizontal, 20)
                .padding(.top, 14)
                .padding(.bottom, 20)
                .readableWidth()
            }
        }
    }

    private var revenueLabel: String {
        switch period {
        case .day: L("dash_revenue_label_day")
        case .week: L("dash_revenue_label_week")
        case .month: L("dash_revenue_label")
        }
    }

    /// Period-over-period revenue sketch; the demo layer ships no per-day series yet.
    private var revenueBars: some View {
        let heights: [CGFloat] = [0.42, 0.55, 0.48, 0.72, 0.5, 1, 0.6, 0.38]
        return HStack(alignment: .bottom, spacing: 7) {
            ForEach(Array(heights.enumerated()), id: \.offset) { index, fraction in
                RoundedRectangle(cornerRadius: 6)
                    .fill(index == 5 ? Color.oliveWood : Color.bone.opacity(0.12))
                    .frame(maxWidth: .infinity)
                    .frame(height: 44 * fraction)
            }
        }
        .frame(height: 44, alignment: .bottom)
    }

/// Sotto questa soglia un giorno aperto è "vuoto": è lì che una campagna serve.
    private static let emptyDayThreshold = 50

    /// I prossimi 7 giorni: una colonna per giorno con l'occupazione, i giorni
    /// chiusi segnati, chi aspetta in lista ("Avvisami") e, se c'è un giorno
    /// vuoto, la scorciatoia per riempirlo con una campagna.
    private func upcomingSection(_ days: [UpcomingDay]) -> some View {
        let waiting = days.reduce(0) { $0 + $1.waitlistCount }
        let hasEmptyDay = days.contains { !$0.closed && $0.occupancyPercent < Self.emptyDayThreshold }
        return VStack(alignment: .leading, spacing: 10) {
            BrandSectionLabel(text: L("dash_upcoming"))
            HStack(alignment: .bottom, spacing: 6) {
                ForEach(days) { day in
                    upcomingColumn(day)
                }
            }
            if waiting > 0 {
                Text(L("dash_upcoming_waitlist", waiting))
                    .font(Typo.jost(12))
                    .foregroundStyle(Color.textMuted)
            }
            // L'unico ingresso alla campagna push: sempre presente, e dice "riempi"
            // quando c'è davvero un giorno vuoto.
            AccentButton(
                text: L(hasEmptyDay ? "dash_upcoming_fill" : "dash_send_campaign"),
                action: onSendCampaign,
                height: 52,
                corner: 16,
                leadingSystemImage: "bell"
            )
            .padding(.top, 4)
        }
    }

    private func upcomingColumn(_ day: UpcomingDay) -> some View {
        let parts = formatDateShort(day.date).split(separator: " ")
        let barHeight: CGFloat = 56
        return VStack(spacing: 5) {
            // Chi aspetta un posto quel giorno: la domanda che l'agenda non ha servito.
            Text(day.waitlistCount > 0 ? "\(day.waitlistCount)" : " ")
                .font(Typo.jost(11, weight: .medium))
                .foregroundStyle(Color.bone)
                .frame(minWidth: 18, minHeight: 18)
                .background(Circle().fill(day.waitlistCount > 0 ? Color.oliveWood : Color.clear))
            ZStack(alignment: .bottom) {
                RoundedRectangle(cornerRadius: 6).fill(day.closed ? Color.stoneSoft : Color.stone)
                if !day.closed {
                    RoundedRectangle(cornerRadius: 6)
                        .fill(day.occupancyPercent >= 80 ? Color.oliveWood : Color.ink)
                        .frame(height: barHeight * CGFloat(day.occupancyPercent) / 100)
                }
            }
            .frame(height: barHeight)
            Text(day.closed ? L("dash_upcoming_closed") : "\(day.occupancyPercent)%")
                .font(Typo.jost(11, weight: .medium))
                .foregroundStyle(day.closed ? Color.textMuted : Color.ink)
                .lineLimit(1)
                .minimumScaleFactor(0.8)
            VStack(spacing: 0) {
                Text(parts.first.map { String($0).uppercased() } ?? "")
                    .font(Typo.jost(11, weight: .medium))
                    .kerning(0.8)
                    .foregroundStyle(Color.textMuted)
                Text(parts.count > 1 ? String(parts[1]) : "")
                    .font(Typo.cormorant(16, weight: .regular))
                    .foregroundStyle(Color.ink)
            }
        }
        .frame(maxWidth: .infinity)
    }

}

File: android/feature/admin/src/main/kotlin/com/devora/mencare/feature/admin/dashboard/DashboardScreen.kt
package com.devora.mencare.feature.admin.dashboard

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.NotificationsNone
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.em
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.devora.mencare.core.common.formatDateShort
import com.devora.mencare.core.common.formatPriceCompact
import com.devora.mencare.core.designsystem.component.AccentButton
import com.devora.mencare.core.designsystem.component.BrandSectionLabel
import com.devora.mencare.core.designsystem.component.DarkHeader
import com.devora.mencare.core.designsystem.component.SegmentedTabs
import com.devora.mencare.core.designsystem.component.StatTile
import com.devora.mencare.core.designsystem.component.readableWidth
import com.devora.mencare.core.designsystem.theme.Bone
import com.devora.mencare.core.designsystem.theme.Cormorant
import com.devora.mencare.core.designsystem.theme.Ink
import com.devora.mencare.core.designsystem.theme.OliveLight
import com.devora.mencare.core.designsystem.theme.OliveWood
import com.devora.mencare.core.designsystem.theme.Overline
import com.devora.mencare.core.designsystem.theme.Stone
import com.devora.mencare.core.designsystem.theme.StoneBorder
import com.devora.mencare.core.designsystem.theme.StoneSoft
import com.devora.mencare.core.designsystem.theme.TextMuted
import com.devora.mencare.core.model.DashboardPeriod
import com.devora.mencare.core.model.UpcomingDay
import com.devora.mencare.feature.admin.R

@Composable
fun DashboardScreen(
    onSendCampaign: () -> Unit,
    viewModel: DashboardViewModel = hiltViewModel(),
) {
    val state by viewModel.state.collectAsStateWithLifecycle()
    val stats = state.stats

    Column(modifier = Modifier.fillMaxSize().background(Bone).verticalScroll(rememberScrollState())) {
        DarkHeader(
            roundedBottom = true,
            contentPadding = PaddingValues(start = 20.dp, end = 20.dp, top = 8.dp, bottom = 18.dp),
        ) {
            // La banda porta il suo titolo, come le altre schermate del
            // titolare: prima si sa dove si è, poi si sceglie il periodo.
            Text(
                stringResource(R.string.dash_overline).uppercase(),
                style = Overline,
                color = OliveLight,
            )
            Spacer(Modifier.height(4.dp))
            Text(
                stringResource(R.string.dash_title),
                style = MaterialTheme.typography.displaySmall,
                color = Bone,
            )
            Spacer(Modifier.height(16.dp))
            SegmentedTabs(
                options = listOf(
                    stringResource(R.string.dash_period_day),
                    stringResource(R.string.dash_period_week),
                    stringResource(R.string.dash_period_month),
                ),
                selectedIndex = state.period.ordinal,
                onSelect = { viewModel.setPeriod(DashboardPeriod.entries[it]) },
            )
            Spacer(Modifier.height(16.dp))
            Text(
                stringResource(
                    when (state.period) {
                        DashboardPeriod.DAY -> R.string.dash_revenue_label_day
                        DashboardPeriod.WEEK -> R.string.dash_revenue_label_week
                        DashboardPeriod.MONTH -> R.string.dash_revenue_label
                    },
                ).uppercase(),
                style = MaterialTheme.typography.labelSmall.copy(fontSize = 11.sp, letterSpacing = 0.16.em),
                color = OliveWood,
            )
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(
                    formatPriceCompact(stats?.revenueCents ?: 0),
                    fontFamily = Cormorant,
                    fontSize = 44.sp,
                    color = Bone,
                    modifier = Modifier.weight(1f),
                )
                Text(
                    stringResource(R.string.dash_trend, stats?.revenueTrendPercent ?: 0),
                    style = MaterialTheme.typography.titleSmall,
                    color = Bone,
                    modifier = Modifier
                        .clip(RoundedCornerShape(10.dp))
                        .background(OliveWood)
                        .padding(horizontal = 12.dp, vertical = 7.dp),
                )
            }
            Spacer(Modifier.height(10.dp))
            RevenueBars()
        }

        Column(
            modifier = Modifier.readableWidth().padding(horizontal = 20.dp).padding(top = 14.dp, bottom = 20.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            Row(
                modifier = Modifier.height(IntrinsicSize.Min),
                horizontalArrangement = Arrangement.spacedBy(10.dp),
            ) {
                // I tre KPI hanno lo stesso aspetto: fondo chiaro e filo, niente grigio.
                StatTile(
                    label = stringResource(R.string.dash_appointments),
                    value = "${stats?.appointmentCount ?: 0}",
                    modifier = Modifier.weight(1f).fillMaxHeight(),
                    container = Bone,
                    borderColor = StoneBorder,
                )
                StatTile(
                    label = stringResource(R.string.dash_no_show),
                    value = String.format(java.util.Locale.ITALIAN, "%.1f%%", stats?.noShowPercent ?: 0.0),
                    modifier = Modifier.weight(1f).fillMaxHeight(),
                    container = Bone,
                    borderColor = StoneBorder,
                )
                StatTile(
                    label = stringResource(R.string.dash_avg_ticket),
                    value = formatPriceCompact(stats?.averageTicketCents ?: 0),
                    modifier = Modifier.weight(1f).fillMaxHeight(),
                    container = Bone,
                    borderColor = StoneBorder,
                )
            }

            Column {
                BrandSectionLabel(stringResource(R.string.dash_operator_occupancy))
                Spacer(Modifier.height(9.dp))
                stats?.operatorOccupancy?.forEach { row ->
                    Row(
                        modifier = Modifier.fillMaxWidth().padding(bottom = 9.dp),
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        Text(
                            row.operatorName.substringBefore(' '),
                            style = MaterialTheme.typography.bodyMedium,
                            color = Ink,
                            modifier = Modifier.width(64.dp),
                        )
                        Box(
                            modifier = Modifier
                                .weight(1f)
                                .height(7.dp)
                                .clip(RoundedCornerShape(4.dp))
                                .background(Stone),
                        ) {
                            Box(
                                modifier = Modifier
                                    .fillMaxWidth(row.percent / 100f)
                                    .height(7.dp)
                                    .clip(RoundedCornerShape(4.dp))
                                    .background(if (row.percent >= 80) OliveWood else Ink),
                            )
                        }
                        Text(
                            "${row.percent}%",
                            style = MaterialTheme.typography.titleSmall,
                            color = Ink,
                            modifier = Modifier.width(44.dp).padding(start = 10.dp),
                        )
                    }
                }
            }

            UpcomingSection(stats?.upcomingDays.orEmpty(), onSendCampaign)
        }
    }
}

/** Period-over-period revenue sketch; the demo layer ships no per-day series yet. */
@Composable
private fun RevenueBars() {
    val heights = listOf(0.42f, 0.55f, 0.48f, 0.72f, 0.5f, 1f, 0.6f, 0.38f)
    Row(
        modifier = Modifier.fillMaxWidth().height(44.dp),
        horizontalArrangement = Arrangement.spacedBy(7.dp),
        verticalAlignment = Alignment.Bottom,
    ) {
        heights.forEachIndexed { index, fraction ->
            Box(
                modifier = Modifier
                    .weight(1f)
                    .fillMaxHeight(fraction)
                    .clip(RoundedCornerShape(6.dp))
                    .background(if (index == 5) OliveWood else Bone.copy(alpha = 0.12f)),
            )
        }
    }
}

/** Sotto questa soglia un giorno aperto è "vuoto": è lì che una campagna serve. */
private const val EMPTY_DAY_THRESHOLD = 50

/**
 * I prossimi 7 giorni: una colonna per giorno con l'occupazione, i giorni
 * chiusi segnati, chi aspetta in lista ("Avvisami") e, se c'è un giorno vuoto,
 * la scorciatoia per riempirlo con una campagna.
 */
@Composable
private fun UpcomingSection(days: List<UpcomingDay>, onSendCampaign: () -> Unit) {
    val waiting = days.sumOf { it.waitlistCount }
    val hasEmptyDay = days.any { !it.closed && it.occupancyPercent < EMPTY_DAY_THRESHOLD }
    Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
        BrandSectionLabel(stringResource(R.string.dash_upcoming))
        Row(
            horizontalArrangement = Arrangement.spacedBy(6.dp),
            verticalAlignment = Alignment.Bottom,
        ) {
            days.forEach { day -> UpcomingColumn(day, Modifier.weight(1f)) }
        }
        if (waiting > 0) {
            Text(
                stringResource(R.string.dash_upcoming_waitlist, waiting),
                style = MaterialTheme.typography.bodySmall.copy(fontSize = 12.sp),
                color = TextMuted,
            )
        }
        // L'unico ingresso alla campagna push: sempre presente, e dice "riempi"
        // quando c'è davvero un giorno vuoto.
        AccentButton(
            text = stringResource(if (hasEmptyDay) R.string.dash_upcoming_fill else R.string.dash_send_campaign),
            onClick = onSendCampaign,
            height = 52.dp,
            shape = RoundedCornerShape(16.dp),
            leadingIcon = Icons.Outlined.NotificationsNone,
            modifier = Modifier.padding(top = 4.dp),
        )
    }
}

@Composable
private fun UpcomingColumn(day: UpcomingDay, modifier: Modifier) {
    val parts = formatDateShort(day.date).split(" ")
    val barHeight = 56.dp
    Column(modifier = modifier, horizontalAlignment = Alignment.CenterHorizontally) {
        // Chi aspetta un posto quel giorno: la domanda che l'agenda non ha servito.
        Box(
            modifier = Modifier
                .size(18.dp)
                .clip(CircleShape)
                .background(if (day.waitlistCount > 0) OliveWood else Color.Transparent),
            contentAlignment = Alignment.Center,
        ) {
            if (day.waitlistCount > 0) {
                Text(
                    "${day.waitlistCount}",
                    style = MaterialTheme.typography.labelSmall.copy(fontSize = 11.sp, letterSpacing = 0.sp),
                    color = Bone,
                )
            }
        }
        Spacer(Modifier.height(5.dp))
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(barHeight)
                .clip(RoundedCornerShape(6.dp))
                .background(if (day.closed) StoneSoft else Stone),
            contentAlignment = Alignment.BottomCenter,
        ) {
            if (!day.closed) {
                Box(
                    Modifier
                        .fillMaxWidth()
                        .height(barHeight * (day.occupancyPercent / 100f))
                        .clip(RoundedCornerShape(6.dp))
                        .background(if (day.occupancyPercent >= 80) OliveWood else Ink),
                )
            }
        }
        Spacer(Modifier.height(5.dp))
        Text(
            if (day.closed) stringResource(R.string.dash_upcoming_closed) else "${day.occupancyPercent}%",
            style = MaterialTheme.typography.labelSmall.copy(fontSize = 11.sp, letterSpacing = 0.sp),
            color = if (day.closed) TextMuted else Ink,
            maxLines = 1,
        )
        Spacer(Modifier.height(5.dp))
        Text(
            parts.getOrElse(0) { "" }.uppercase(),
            style = MaterialTheme.typography.labelSmall.copy(fontSize = 11.sp, letterSpacing = 0.08.em),
            color = TextMuted,
        )
        Text(
            parts.getOrElse(1) { "" },
            fontFamily = Cormorant,
            fontSize = 16.sp,
            color = Ink,
        )
    }
}
